docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,414 @@
+simply need to call the right program to perform instructions.
+
+## Required Changes
+
+A new interest-bearing token program is only one component of the overall solution,
+which also comprises an interface for tokens (i-tokens) to be easily integrated in
+on-chain programs, web apps, and wallets.
+
+### Token Program Conformance
+
+Since our solution entails the creation of a new token program, we're opening the
+floodgates for other app developers to do the same. To ensure safe and easy integration
+of new token programs, we need a set of conformance tests based on the current
+SPL token test suite.
+
+#### Instructions
+
+A new token program must include all of the same instructions as SPL token,
+but the "unchecked" variants should return errors. Concretely, this means the
+program must implement:
+
+* `InitializeMint`
+* `InitializeHolding` (previously `InitializeAccount`)
+* `Revoke`
+* `SetAuthority`
+* `CloseHolding` (previously `CloseAccount`)
+* `FreezeHolding` (previously `FreezeAccount`)
+* `ThawHolding` (previously `ThawAccount`)
+* `TransferChecked`
+* `ApproveChecked`
+* `MintToChecked`
+* `BurnChecked`
+* `InitializeHoldingWithOwnerAddress` (previously `InitializeAccount2`)
+
+And the program should throw errors for:
+
+* `Transfer`
+* `Approve`
+* `MintTo`
+* `Burn`
+* `InitializeMultisig` (use a more general multisig)
+
+New instructions required:
+
+* `CreateHolding`: only performs `Allocate` and `Assign` to self,
+useful for creating a holding when you don't know the program that you're
+interacting with. See the Associated Token Account section for how to use this.
+
+There are also new read-only instructions, which write data to a provided account
+buffer:
+
+* `AmountToUiAmount`: convert the given share amount to a UI token amount
+* `UiAmountToAmount`: convert the given UI token amount to an exact share amount
+
+See the Runtime section for more information on how these are used.
+
+Programs and wallets that wish to support i-tokens must update the instructions
+to use all `Checked` variants and use the new read-only instructions to fetch
+information.
+
+#### Conversions and Wrapper Library
+
+The SPL token library provides wrappers compatible with any conforming program.
+The wrapper simply checks the program id and deserializes the account data
+as an `spl_token::Holding` or `spl_token::Mint`. If a client program requires access
+to data specific to an i-token program, it needs to use that program's serializers.
+
+To convert between amount and UI amount, the wrapper library calls the `AmountToUiAmount`
+instruction on the i-token program, or for SPL token accounts, it calls
+`spl_token::amount_to_ui_amount` inline.
+
+For interest-bearing tokens, after deserialization into a `spl_token::Holding`,
+the `amount` field will be in terms of shares, and not token amount, so UIs and
+on-chain programs that need token amounts must convert using `AmountToUiAmount`.
```

Comment:
stale paragraph?

Reply:
Michael and I had a long discussion about this part, so I hope it isn't stale :laughing:

We ended up deciding that `amount` is in terms of shares, so dapps / programs need to be aware that `ui_amount` can potentially be totally different due to a calculation during the `AmountToUiAmount` conversion, and know explicitly when to use `amount` vs `ui_amount`.  This makes interest-bearing tokens look just like normal tokens.  We still need some concrete feedback from the issuers of interest-bearing tokens to make sure this works for their system, but it also works much more easily for bridges.

Reply:
I see! thanks for the clarification.